File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        int[] parents = new int[N + 1];
        boolean[] bVisited = new boolean[N + 1];

        ArrayList<Integer>[] graph = new ArrayList[N + 1];
        for (int i = 0; i <= N; i++) {
            graph[i] = new ArrayList<>();
        }

        int src, dest;
        StringTokenizer st;
        for (int i = 0; i < N - 1; i++) {
            st = new StringTokenizer(br.readLine());
            src = Integer.parseInt(st.nextToken());
            dest = Integer.parseInt(st.nextToken());

            graph[src].add(dest);
            graph[dest].add(src);
        }

        Queue<Integer> Q = new LinkedList<>();
        Q.add(1);
        bVisited[1] = true;

        int cur;
        while (!Q.isEmpty()) {
            cur = Q.poll();
            bVisited[cur] = true;

            for (int child : graph[cur]) {
                if (!bVisited[child]) {
                    parents[child] = cur;
                    Q.add(child);
                }
            }
        }

        for (int i = 2; i <= N; i++) {
            System.out.println(parents[i]);
        }
    }
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.cc
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

int N;
vector<int> Graph[1000001];
int Parents[1000001];
int bVisited[1000001];

int main()
{
	cin >> N;
	
	int src, dest;
	for (int i = 0; i < N - 1; i++)
	{
		cin >> src >> dest;
		Graph[src].push_back(dest);
		Graph[dest].push_back(src);
	}

	queue<int> Q;
	Q.push(1);
	bVisited[1] = true;

	int cur;
	while (!Q.empty())
	{
		cur = Q.front();
		Q.pop();
		bVisited[cur] = true;

		for (int next : Graph[cur])
		{
			if (!bVisited[next])
			{
				Q.push(next);
				Parents[next] = cur;
			}
		}
	}

	for (int i = 2; i <= N; i++)
	{
		cout << Parents[i] << '\n';
	}

	return 0;
}